### PR TITLE
Add Service SLO Metrics

### PR DESCRIPTION
**Pull request by violetyao**
Jira: https://issues.redhat.com/browse/APPSRE-2955
Example usage:
```
$ curl --request POST \
--header "Content-Type: application/json" \
--data @examples/serviceslometrics.json \
localhost:8080/api/v1/serviceslometrics/app-sre-prod-04  

"ok"
```

```
$ curl "localhost:8080/api/v1/serviceslometrics?cluster=app-sre-prod-04&namespace=cincinnati-production&sli_type=latency&name=policy-engine-latency"  

{
  "cluster": "app-sre-prod-04",
  "name": "policy-engine-latency",
  "namespace": "cincinnati-production",
  "service": "cincinnati",
  "sli_type": "availability",
  "target": 99,
  "value": 66
}
```

```
$ curl "localhost:8080/api/v1/metrics"

# HELP imagemanifestvuln_total Vulnerabilities total per severity
# TYPE imagemanifestvuln_total counter
# HELP deploymentvalidation_total Validations by validation type
# TYPE deploymentvalidation_total counter
# HELP serviceslometrics ServiceSLOMetrics by cluster, namespace, slitype, name
# TYPE serviceslometrics gauge
# HELP imagemanifestvuln_total Vulnerabilities total per severity
# TYPE imagemanifestvuln_total counter
# HELP deploymentvalidation_total Validations by validation type
# TYPE deploymentvalidation_total counter
# HELP serviceslometrics ServiceSLOMetrics by cluster,namespace, slitype, name, type
# TYPE serviceslometrics gauge
serviceslometrics{cluster="app-sre-prod-04",name="policy-engine latency",namespace="cincinnati-production",slitype="latency",type="slo_value"} 99.0
serviceslometrics{cluster="app-sre-prod-04",name="policy-engine latency",namespace="cincinnati-production",slitype="latency",type="slo_target"} 95.0
serviceslometrics{cluster="app-sre-prod-01",name="Reconciliation time of app-interface integrations",namespace="app-interface-production",slitype="latency",type="slo_value"} 76.0
serviceslometrics{cluster="app-sre-prod-01",name="Reconciliation time of app-interface integrations",namespace="app-interface-production",slitype="latency",type="slo_target"} 95.0
serviceslometrics{cluster="app-sre-prod-04",name="latency",namespace="uhc-production",slitype="latency",type="slo_value"} 99.0
serviceslometrics{cluster="app-sre-prod-04",name="latency",namespace="uhc-production",slitype="latency",type="slo_target"} 95.0
serviceslometrics{cluster="app-sre-prod-04",name="policy-engine errors",namespace="cincinnati-production",slitype="availability",type="slo_value"} 100.0
serviceslometrics{cluster="app-sre-prod-04",name="policy-engine errors",namespace="cincinnati-production",slitype="availability",type="slo_target"} 95.0
serviceslometrics{cluster="app-sre-prod-04",name="policy-engine availability",namespace="cincinnati-production",slitype="availability",type="slo_value"} 99.0
serviceslometrics{cluster="app-sre-prod-04",name="policy-engine availability",namespace="cincinnati-production",slitype="availability",type="slo_target"} 95.0
serviceslometrics{cluster="app-sre-prod-04",name="errors",namespace="uhc-production",slitype="availability",type="slo_value"} 99.0
serviceslometrics{cluster="app-sre-prod-04",name="errors",namespace="uhc-production",slitype="availability",type="slo_target"} 95.0
serviceslometrics{cluster="app-sre-prod-04",name="availability",namespace="uhc-production",slitype="availability",type="slo_value"} 99.0
serviceslometrics{cluster="app-sre-prod-04",name="availability",namespace="uhc-production",slitype="availability",type="slo_target"} 95.0
```
Diagram: 
dashdotdb

### CODE REVIEW

**Comment by violetyao** [comment]
@BumbleFeng PTAL when you have time. ;D Thank you!

**Comment by apahim** [comment]
Do you plan to fix the ItemsDeploymentValidation or should I look into that?

**Comment by apahim** [comment]
While on it, can you please remove those blue annotation from the ERD? They have no reason to exist...

File: dashdotdb/controllers/serviceslometrics.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+def post(cluster, body):
+    imv = ServiceSLOMetrics(cluster=cluster)
+    imv.insert(slo=body)
+    return 'ok'
+
+
+def search(cluster, namespace, sli_type, name):
+    imv = ServiceSLOMetrics(cluster, namespace, sli_type, name)
+    return imv.get_slometrics()
```
Comment (apahim) [inline dashdotdb/controllers/serviceslometrics.py:12]: Are you also going to expose condensed data in the /metrics endpoint?

**Comment by apahim** [comment]
What a beatigul PR. I'm impressed :)

**Comment by violetyao** [comment]
> Do you plan to fix the ItemsDeploymentValidation or should I look into that?

I can take a look!

**Comment by violetyao** [comment]
> While on it, can you please remove those blue annotation from the ERD? They have no reason to exist...

Do we keep those  1 -> n arrows or shall we remove both 1 -> n arrows and the blue annotations (e.g 1 {deployment validation}{id})?

File: dashdotdb/controllers/serviceslometrics.py (hunk 1 above)
Comment (violetyao) [reply dashdotdb/controllers/serviceslometrics.py:12]: Good point! I'll do that. ;D

File: dashdotdb/models/dashdotdb.py [hunk 2]
```diff
@@ -131,3 +143,27 @@ class ObjectKind(db.Model):
...
+    __tablename__ = 'serviceslo'
+
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(64), unique=False)
+    value = db.Column(db.Integer, unique=False)
+    target = db.Column(db.Integer, unique=False)
+    slitype_id = db.Column(db.Integer, db.ForeignKey('slitype.id'))
+    token_id = db.Column(db.Integer, db.ForeignKey('token.id'))
+    service_id = db.Column(db.Integer, db.ForeignKey('service.id'))
+    cluster_id = db.Column(db.Integer, db.ForeignKey('cluster.id'))
```
Comment (apahim) [inline dashdotdb/models/dashdotdb.py:168]: Sorry, missed this in the first round. No need for the cluster_id here. You get to the cluster via the namespace.

File: dashdotdb/schemas/swagger.yaml [hunk 3]
```diff
@@ -296,5 +411,6 @@ components:
                   value:
                     type: array
                     items:
-                    - type: number
-                    - type: string
+                      oneOf:
+                        - type: number
+                        - type: string
```
Comment (apahim) [inline dashdotdb/schemas/swagger.yaml:416]: Thank you!

**Comment by apahim** [comment]
Thank for the updates. There's a small issue with the new table and that will change a bot your queries. With that, we will be ready to go :)

**Comment by apahim** [comment]
Remember to update the ERD, dropping the direct link between Cluster and ServiceSLOMetrics.

File: dashdotdb/models/dashdotdb.py [hunk 4]
```diff
@@ -131,3 +143,27 @@ class ObjectKind(db.Model):
...
+class SLIType(db.Model):
+
+    __tablename__ = 'slitype'
+
+    id = db.Column(db.Integer, primary_key=True)
+    name = db.Column(db.String(64), unique=False)
+    serviceslo = db.relationship('ServiceSLO', backref='slitype')
+
+
+class ServiceSLO(db.Model):
```
Comment (apahim) [inline dashdotdb/models/dashdotdb.py:157]: The ERD shows a different table name.

Comment (apahim) [reply dashdotdb/models/dashdotdb.py:157]: I like this name better btw.

File: dashdotdb/models/dashdotdb.py (hunk 2 above)
Comment (violetyao) [reply dashdotdb/models/dashdotdb.py]: Gotcha!

File: dashdotdb/models/dashdotdb.py (hunk 4 above)
Comment (violetyao) [reply dashdotdb/models/dashdotdb.py:156]: haha i see! Thank you. ;D

**Comment by apahim** [comment]
Oh, there's one missing link in the ERD between Token and ServiceSLO

**Comment by violetyao** [comment]
Oops my bad! Updated. Thanks for spotting that. :)

**Comment by violetyao** [comment]
@BumbleFeng Hi Feng! I've updated controller/metrics.py so that slo_gauge's values are now reflecting slo_target/ slo_value. Let me know what you think! Thank you.;)

**Comment by BumbleFeng** [comment]
LGTM

**Pull request merged**